src/TestableIO.System.IO.Abstractions.TestingHelpers/Events/MockFileSystemEvents.cs

```diff
@@ -0,0 +1,394 @@
+    /// </summary>
+    /// <typeparam name="T">The return type of the operation</typeparam>
+    /// <param name="path">The file path</param>
+    /// <param name="operation">The type of operation</param>
+    /// <param name="resourceType">The type of resource</param>
+    /// <param name="func">The operation to execute</param>
+    /// <returns>The result of the operation</returns>
+    public T WithEvents<T>(string path, FileOperation operation, ResourceType resourceType, Func<T> func)
+    {
+        if (!isEnabled)
+        {
+            return func();
+        }
+            
+        RaiseOperation(path, operation, resourceType, OperationPhase.Before);
+        
+        var operationCompleted = false;
+        try
+        {
+            var result = func();
+            operationCompleted = true;
+            return result;
+        }
+        finally
+        {
+            if (operationCompleted)
+            {
+                try
+                {
+                    RaiseOperation(path, operation, resourceType, OperationPhase.After);
+                }
+                catch
+                {
+                    // Don't let After event exceptions mask the main operation
+                }
+            }
+        }
+    }
+    
+    /// <summary>
+    /// Wraps a void operation with Before/After events, handling exceptions properly.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <param name="operation">The type of operation</param>
+    /// <param name="resourceType">The type of resource</param>
+    /// <param name="action">The operation to execute</param>
+    public void WithEvents(string path, FileOperation operation, ResourceType resourceType, Action action)
+    {
+        if (!isEnabled)
+        {
+            action();
+            return;
+        }
+            
+        RaiseOperation(path, operation, resourceType, OperationPhase.Before);
+        
+        var operationCompleted = false;
+        try
+        {
+            action();
+            operationCompleted = true;
+        }
+        finally
+        {
+            if (operationCompleted)
+            {
+                try
+                {
+                    RaiseOperation(path, operation, resourceType, OperationPhase.After);
+                }
+                catch
+                {
+                    // Don't let After event exceptions mask the main operation
+                }
+            }
+        }
+    }
+    
+    /// <summary>
+    /// Raises an operation event if events are enabled.
+    /// </summary>
+    internal void RaiseOperation(
+        string path, 
+        FileOperation operation, 
+        ResourceType resourceType, 
+        OperationPhase phase)
+    {
+        if (!isEnabled) {
+            return;
+        }
+            
+        var args = new FileSystemOperationEventArgs(path, operation, resourceType, phase);
+        Subscription[] currentSubscriptions;
```

Comment:
For thread-safe collections you should be able to directly iterate over the `subscriptions` and can avoid introducing a temporary copy of the collection.